sklearn/tests/test_array_out.py

```diff
@@ -0,0 +1,27 @@
+import pytest
+
+import numpy as np
+from numpy.testing import assert_array_equal
+
+from sklearn.utils._array_out import _get_feature_names
+
+
+def _construct_array(array_type, column_names):
+    X = np.array([[1, 2, 3], [4, 5, 6]], dtype=float)
+
+    if array_type == "dataframe":
+        pd = pytest.importorskip("pandas")
+        return pd.DataFrame(X, columns=column_names)
+    elif array_type == "dataarray":
+        xr = pytest.importorskip("xarray")
+        return xr.DataArray(X, dims=('index', 'columns'),
+                            coords={'columns': column_names})
+
+
+@pytest.mark.parametrize("array_type", ["dataframe", "dataarray"])
+def test_pandas_get_feature_names(array_type):
```

Comment:
can you test with integer column names as well? and maybe a mix of integer, string and object column names? you know, for fun?

Reply:
I remember we had a conversation about what kind of column names we'd accept. And we kinda concluded that we'd restrict column names to be `str`, and I think we should stick to that?

Reply:
I would be okay with this if all our `get_feature_names` output strings, but for `DictVectorizer` we do not:

```py
from sklearn.feature_extraction import DictVectorizer
v = DictVectorizer(sparse=False)
X = [{2: 1, 4: 2}, {3: 3, 5: 1}]
v.fit(X)
v.get_feature_names()
# [2, 3, 4, 5]
```

On a second note, I think enforcing strings will break backward compatibly: 

```py
from sklearn.preprocessing import StandardScaler
from sklearn.datasets import load_iris

X, _ = load_iris(as_frame=True, return_X_y=True)
X.columns = ['a', 'b', 2, 3]
scaler = StandardScaler().fit(X)
_ = scaler.transform(X) # currently works
```

Reply:
I know integers as columns currently works, but I think we should deprecate the usage and start returning strings as column names ourselves. Having integers as column names (or row indices for that matter) creates a ton of silent bugs.

Reply:
Becoming consistent would be great and since `ColumnTransformer` is also not accepting mixed data type (and I am not sure that we can get around that without adding some ambuiguities), I would prefer to deprecate the mixed type support where it is currently.

Reply:
Another option would be to accept integer columns in `_validate_data` and automatically convert them to be stored as strings in `feature_names_in_`.

I am not sure this is a good idea though.

Reply:
To summarize here are the options:

1. Only store `feature_names_in_` when column names are all strings. If a dataframe is passed with mixed or integer column names, show a warning stating that `feature_names_in_` is not set.
2. Support integer and/or mixed columns and deprecate.
3. Convert to strings and store them into `feature_names_in_` and warn that this is happening.
4. Support any type of column names.

Option 1 allows us to go in either direction in the future without breaking BC. I am +0.5 for Option 2 as long as we update `DictVectorizer` in a way to always output strings. Option 3 will lead to weird edge cases like column names: `['4', 4]` (string and integer).

Option 4 is not consistent with `ColumnTransformer`. `ColumnTransformer` uses the names to actually slice the DataFrame so its reasonable to enforce strings. All the other estimators only requires an equality check, so I do not think we need to be restrictive on the column names.

To move this along I'll update the PR to option 2.

Reply:
Do we need a SLEP or just a vote on how to handle column names? I think some of us would be happy with whatever pandas accepts, and some of us [really] don't want any column names other than simple strings.

Reply:
How we handle names is fairly public, so having a short SLEP to formalize the behavior is useful.